chips/sam4l/src/usart.rs

```diff
@@ -716,6 +724,17 @@ impl USART {
         // enable timeout interrupt
         usart.registers.ier.write(Interrupt::TIMEOUT::SET);
 
+        // Have to re-enable the NVIC interrupt for this USART after enabling
+        // the `TIMEOUT` interrupt for some inexplicable reason. This makes
+        // entirely no sense, but after testing `receive_automatic` for nRF
+        // serialization and using AT commands, across multiple boards and
+        // USARTs, something about enabling the TIMEOUT interrupt above causes
+        // the USART NVIC interrupt to become disabled.
+        unsafe {
+            use cortexm4;
+            cortexm4::nvic::Nvic::new(self.nvic_index).enable();
+        }
+
```

Comment:
Here's my guess for what's happening:

  1. Line 725 enables the TIMEOUT interrupt

  2. (For some reason) the interrupt fires immediately, triggering `generic_isr`

  3. `generic_isr` disables the NVIC, since that's what it does. At this point the NVIC is still pending, though.

  4. Line 230 (now 739) starts the Timeout which, according to the datasheet, implicitly clears the TIMEOUT interrupt bit in CSR: "Writing a one to this bit will abort any current time-out count down, and trigger a new count down when the next character has been received. CSR.TIMEOUT is also cleared."

  5. At this point, the NVIC pending bit clears, since the interrupt is no longer valid, and the NVIC is still disabled, since `service_pending_interrupts` has not yet executed.

  6. Now the chip goes to sleep. When this interrupt fires, it will not wake up the CPU since the NVIC is disabled, though the NVIC _will_ become pending.

  7. The CPU wakes up at some later point due to a different event. Now when `service_pending_interrupts` runs, it'll still catch the timeout event since the NVIC bit is pending.

I think the remedy might be simply to move the start timeout command to before enabling the interrupt (so swap lines 720 and 725 in the original source).

Reply:
I think you are exactly right. PR fixed. Hooray!